scene/main/canvas_item.cpp

```diff
@@ -557,37 +557,41 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
 		}
 
 		RenderingServer::get_singleton()->canvas_item_add_rect(canvas_item, p_rect, p_color);
+	} else if (p_width >= p_rect.size.width || p_width >= p_rect.size.height) {
+		RenderingServer::get_singleton()->canvas_item_add_rect(canvas_item, p_rect.grow(0.5f * p_width), p_color);
 	} else {
-		// Thick lines are offset depending on their width to avoid partial overlapping.
-		// Thin lines don't require an offset, so don't apply one in this case
-		real_t offset;
-		if (p_width >= 2) {
-			offset = p_width / 2.0;
-		} else {
-			offset = 0.0;
+		// Workaround for scaling issue. See `RendererCanvasCull::canvas_item_add_line()`.
+		if (p_width <= 1.001) {
+			p_width = 1.002;
 		}
```

Comment:
@dalexeev I kinda understand where this change is coming from but I don't think it's acceptable to be merged. Besides it not really fixing the scaling issue (it indeed makes such rects be always scaled but they'd still end up being wider than desired), it also breaks all the cases where drawing a rect with line primitive was desired (e.g. probably most `draw_rect` calls within the built-in GUI controls for drawing unfilled rect with undefined line width (hence using the default 1.0)).

Reply:
If there was a way
choose the desired `draw_line` option (2 point primitive vs 4 point primitive), then this hack would not be needed.

For the "thin rectangle" variant (which doesn't change line thickness when scaled), we have to consider the CanvasItem's transform?

I think one possible solution to this dilemma is to separate the API, like so:

* `draw_line` - always draws a 4 point primitive.
* `draw_thin_line` - draws a 2 point primitive.

Reply:
> If there was a way
> choose the desired `draw_line` option (2 point primitive vs 4 point primitive), then this hack would not be needed.

Indeed, that's the problem. I've already mentioned it in https://github.com/godotengine/godot/pull/41239#issuecomment-1335792985) (and suggested another possible solution). If going for separating the API (again, can't tell if that's the preffered way of solving it) then I think it'd make more sense to keep the same prefix in both methods for easier discoverability, e.g. something like `draw_line`+`draw_line_primitive` / `draw_line_quad`+`draw_line` / `draw_line_quad`+`draw_line_primitive` / etc.

>For the "thin rectangle" variant (which doesn't change line thickness when scaled), we have to consider the CanvasItem's transform?

If we'd want to keep auto-detecting when to use line primitive over the quad and make it be done only when the actual result is thin (instead of when the rect is locally thin like it's currently done) then yeah, all the transforms which will affect the final CanvasItem render would need to be taken into account. But I don't think it's easily feasible as the decision which primitive to use (line/quad) would need to be probably postponed to when the RenderingServer actually issues stuff to be rendered. And actually I think auto detection might be not wanted as it might be break batching. It seems that primitives with different point count (like lines and quads) are currently not batched together. Which means using always quads could be potentially more performant than sometimes using line primitives. So yeah, I still think just giving the choice for quad/line to the user is the way to go.

Reply:
```suggestion
		// Don't offset if line primitives instead of quads. See `RendererCanvasCull::canvas_item_add_line()`.
		if (p_width <= 1.001) {
			offset = 0.0;
		}
```